yjit/src/asm/arm64/arg/fixed_width.rs

```diff
@@ -0,0 +1,57 @@
+// There are many instances in AArch64 instruction encoding where you represent
+// an integer value with a particular bit width that isn't a power of 2. These
+// structs represent the signed and unsigned versions of this kind of encoding,
+// and provide Into<u32> to allow the individual instruction encoding functions
+// to use them easily.
+
+/// Shorten a signed immediate to fit into a compile-time known width.
+pub fn fix_signed_immediate<T: Into<i32>, const WIDTH: usize>(value: T) -> u32 {
+    let value: i32 = value.into();
+    (value as u32) & ((1 << WIDTH) - 1)
+}
+
+/// Shorten an unsigned immediate to fit into a compile-time known width.
+pub fn fix_unsigned_immediate<T: Into<u32>, const WIDTH: usize>(value: T) -> u32 {
+    let value: u32 = value.into();
+    (value & ((1 << WIDTH) - 1))
+}
```

Comment:
IIUC what these functions are doing is dropping the top bits, truncating integer values.

Maybe it would be better to name them `truncate_imm()` and `truncate_uimm()` ? Currently the name of these functions doesn't clearly indicate that they do IMO.

The other thing is, for an unsigned value, if the input already fits into the required number of bits, this should be a noop. For a signed value that is negative, it will change the value though.

It might be smart to add some assertions to make sure that:
- For an unsigned immediate, the top bits that we're not masking are all zeros (so basically, masked_output_value == input_value)
- For a signed immediate, if the value is positive, we're not dropping any bits, and for a negative value, the masked bits were all ones. There might be a simpler way to write that by saying, if you pad the negative masked value with all ones, you get the input value back.

Reply:
Yeah this makes a lot of sense, thanks for the detailed feedback. I updated per your suggestions.